### PR TITLE
Add reporter for outputting restartable data

### LINKED ISSUE #26304: Add reporter for restartable data
## Reason
Restartable data is often difficult to inspect as it's stored in a binary form. The same goes for mesh meta data. It would be good to have a way to output the data in a human-readable form.

## Design
Add a reporter that can output restartable data (the data that has a to_json spec).

## Impact
New feature.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #26304

### CODE REVIEW

**Comment by loganharbour** [comment]
@lindsayad when you have a moment. Almost all of the line count is updating the json contrib.

This is in support of the data only mesh generation. Lets us easily output restartable (including mesh meta) data for both use and diffing.

**Comment by loganharbour** [comment]
Will be green now.

File: framework/doc/content/source/reporters/MeshMetaDataReporter.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# MeshMetaDataReporter
+
+A Reporter object for outputting mesh metadata in a human-readable form. By default, it executes only on initial.
```
Comment (grmnptr) [inline framework/doc/content/source/reporters/MeshMetaDataReporter.md:3]: Link to mesh metadata page please for those (like myself) who are not familiar with it.

File: framework/doc/content/source/reporters/RestartableDataReporter.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# RestartableDataReporter
+
+A Reporter object for outputting [Restartable.md] data in a human-readable form. This includes
+restartable metadata, such as mesh metadata. This object is particularly useful for inspecting
+restartable data for the purposes of debugging.
```
Comment (grmnptr) [inline framework/doc/content/source/reporters/RestartableDataReporter.md:5]: Very cool feature!

File: framework/doc/content/source/reporters/RestartableDataReporter.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+Only data types that have a `to_json` specialization support having their values output. If the
+[!param](/Reporters/RestartableDataReporter/allow_unimplemented) parameter is set to true,
+restartable data with unimplemented specializations will still have their auxiliary information output
+(type, whether or not the data is loaded/stored, etc).
+
+The [!param](/Reporters/RestartableDataReporter/include) and [!param](/Reporters/RestartableDataReporter/exclude)
+parameters enable the filtering of data by name. With these parameters, you can use `*` as a
```
Comment (grmnptr) [inline framework/doc/content/source/reporters/RestartableDataReporter.md:13]: And it gets even nicer!

File: framework/include/outputs/JsonIO.h [hunk 4]
```diff
@@ -28,4 +37,29 @@ namespace libMesh
...
+  static void to_json(json & j, const std::unique_ptr<T> & v)
+  {
+    if constexpr (std::is_constructible_v<nlohmann::json, T>)
+    {
+      if (v)
+        j = *v;
+      else
+        j = nullptr;
+    }
+    else
```
Comment (grmnptr) [inline framework/include/outputs/JsonIO.h:61]: Do you think we could do a static assert here with `static_assert(Moose::always_false<T>,"This object doesn't have a to_json implementation");` if we are already doing constexpr?

File: framework/include/restart/RestartableData.h [hunk 5]
```diff
@@ -276,4 +320,14 @@ RestartableData<T>::loadInternal(std::istream & stream)
   loadHelper(stream, set(), _context);
 }
 
+template <typename T>
+inline void
+RestartableData<T>::storeJSONValue(nlohmann::json & json) const
+{
+  if constexpr (RestartableData<T>::has_store_json)
+    nlohmann::to_json(json, get());
+  else
```
Comment (grmnptr) [inline framework/include/restart/RestartableData.h:329]: We could throw a compiler error here wit a static assert, what saith you?

File: test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+[]
+
+[Executioner]
+  type = Transient
+  num_steps = 2
+[]
+
+[Reporters/data]
+  type = RestartableDataReporter
+  include = 'FEProblemBase/*/time*'
```
Comment (grmnptr) [inline test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i:18]: I think extending this to a case where you use a vector of includes/excludes might be beneficial to show users that hey this is how you separate the vector in this case. Even though the documentation states that `*` and `?` can be used but it gave me a regexy feeling where `,;` have meaning. Also, maybe emphasizing a little more in the documentation that only these are supported could be nice.

**Review by grmnptr — COMMENTED** [review]
Very cool stuff there. I would still like to ask for another pair of eyes on it considering my limited familiarity with `json`.

File: framework/include/outputs/JsonIO.h (hunk 4 above)
Comment (loganharbour) [reply framework/include/outputs/JsonIO.h:61]: static asserts are applied before constexpr, so it'll always error

File: framework/include/restart/RestartableData.h (hunk 5 above)
Comment (loganharbour) [reply framework/include/restart/RestartableData.h:329]: Nope. This method will still get instantiated for classes that don't have `to_json`, which is totally valid. We just can't call it. We could only do this if we want to require all restartable types to jave a `to_json` spec, which is no bueno.

File: test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i (hunk 6 above)
Comment (loganharbour) [reply test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i:18]: > Also, maybe emphasizing a little more in the documentation that only these are supported could be nice.

How is this not explicit enough? It describes the options

```
The [!param](/Reporters/RestartableDataReporter/include) and [!param](/Reporters/RestartableDataReporter/exclude)
parameters enable the filtering of data by name. With these parameters, you can use `*` as a
wildcard match of any number of characters and `?` as a wildcard match for a single character.
```

I would really prefer not to make the test more complicated for the sake of having an extra line in the documentation include. The test requirements are more important to me than the documentation for an advanced object

File: framework/include/restart/RestartableData.h (hunk 5 above)
Comment (grmnptr) [reply framework/include/restart/RestartableData.h:329]: Okay, resolving then.

File: test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i (hunk 6 above)
Comment (grmnptr) [reply test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i:18]: Okay, I understand your reasoning. I was just looking at this from a naive user's perspective. I am fine with this though, will resolve this.

Comment (loganharbour) [reply test/tests/reporters/restartable_data_reporter/restartable_data_reporter.i:18]: Thanks. I think it's a valid concern - but I don't wanna make the test test more than it needs to to satisfy the requirement.

I would be more concerned if this wasn't considered a more "advanced" object

File: framework/include/outputs/JsonIO.h (hunk 4 above)
Comment (loganharbour) [reply framework/include/outputs/JsonIO.h:61]: Will still get instantiated so we can't do this. See similar comment below in RestartableData.h

**Review by grmnptr — APPROVED** [review]
Great feature!

**Pull request merged**